File: C++/myLeetCode/Linked_List/142. Linked List Cycle II.cpp
// 142. Linked List Cycle II
// https://leetcode.com/problems/linked-list-cycle-ii/

#include "LinkedListUtils.hpp"

ListNode *detectCycle(ListNode *head) {
    if (!head || !head->next) return nullptr;

    ListNode* slow = head;
    ListNode* fast = head;

    while (fast != nullptr && fast->next != nullptr) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) {  // Cycle detected
            fast = head;  // Reset fast to head
            while (slow != fast) {  // Find the entry point of the cycle
                slow = slow->next;
                fast = fast->next;
            }
            return slow;  // Return the entry point of the cycle
        }
    }
    return nullptr;  // No cycle
}

int main() {
    std::vector<int> values = {3, 2, 0, -4};
    ListNode *head = createLinkedList(values); 
    // 设置链表的最后一个节点的 next 指针，形成环
    ListNode *cur = head;
    while (cur->next != nullptr) {
        cur = cur->next;
    }
    cur->next = head->next;

    ListNode *node = detectCycle(head);
    std::cout << node->val << std::endl; // Should print 2

    return 0;
}

File: C#/myLeetCode/Linked_List/142. Linked List Cycle II.cs
// 142. Linked List Cycle II
// https://leetcode.com/problems/linked-list-cycle-ii/

using System;

public class Solution {
    public ListNode DetectCycle(ListNode head) {
        if (head == null || head.next == null) return null;

        ListNode slow = head, fast = head;

        // 使用 Floyd's Cycle Detection 算法
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) break; // cycle detected
        }

        // 检查是否没有循环
        if (fast == null || fast.next == null) return null;

        slow = head;
        while (slow != fast) {
            slow = slow.next;
            fast = fast.next;
        }

        return slow; // 这就是循环开始的节点
    }
}
public class Program {
    public static void Main() {
        // 使用 ListHelpers 创建链表
        int[] nums = {3, 2, 0, -4};
        ListNode head = ListHelpers.ConvertArrayToList(nums);

        // 手动创建循环
        if (head != null && head.next != null && head.next.next != null && head.next.next.next != null) {
            head.next.next.next.next = head.next;
        }

        Solution solution = new Solution();
        ListNode cycleStart = solution.DetectCycle(head);

        if (cycleStart != null) {
            Console.WriteLine($"Cycle detected! Cycle starts at node with value: {cycleStart.val}");
        } else {
            Console.WriteLine("No cycle detected in the linked list.");
        }
    }
}
